packages/contracts/test/utils/assertions.ts

```diff
@@ -104,7 +129,74 @@ export async function expectTransactionFailedAsync(p: sendTransactionResult, rea
     }
     switch (nodeType) {
         case NodeType.Ganache:
-            return expect(p).to.be.rejectedWith(reason);
+            return expect(p).to.be.rejectedWith(
+                new RegExp(`^${constants.GANACHE_REVERT_REASON_PREFIX}${escapeRegExp(reason)}$`),
+            );
+        case NodeType.Geth:
+            logUtils.warn(
+                'WARNING: Geth does not support revert reasons for sendTransaction. This test will pass if the transaction fails for any reason.',
+            );
+            return expectTransactionFailedWithoutReasonAsync(p);
+        default:
+            throw new Error(`Unknown node type: ${nodeType}`);
+    }
+}
+
+/**
+ * Resolves if the the sendTransaction call fails with the given revert reason and decoded revert params.
+ * However, since Geth does not support revert reasons for sendTransaction, this
+ * falls back to expectTransactionFailedWithoutReasonAsync if the backing
+ * Ethereum node is Geth.
+ * @param p a Promise resulting from a sendTransaction call
+ * @param expectedError Expected revert reason with params
+ * @returns a new Promise which will reject if the conditions are not met and
+ * otherwise resolve with no value.
+ */
+export async function expectTransactionFailedWithParamsAsync(
+    p: sendTransactionResult,
+    expectedError: SmartContractError,
+): Promise<void> {
+    // HACK(albrow): This dummy `catch` should not be necessary, but if you
+    // remove it, there is an uncaught exception and the Node process will
+    // forcibly exit. It's possible this is a false positive in
+    // make-promises-safe.
+    p.catch(e => {
+        _.noop(e);
+    });
+
+    if (_.isUndefined(nodeType)) {
+        nodeType = await web3Wrapper.getNodeTypeAsync();
+    }
+    switch (nodeType) {
+        case NodeType.Ganache:
+            try {
+                await p;
+            } catch (err) {
+                if (_.startsWith(err.message, constants.GANACHE_REVERT_REASON_PREFIX)) {
+                    if (err.hashes.length !== 1) {
+                        throw new Error('Expected just one transaction hash in Ganache error object');
+                    }
+                    const txHash = err.hashes[0];
+                    const ganacheError = err.results[txHash];
+                    const returnData: string = ganacheError.return;
+                    const REVERT_ERROR_SELECTOR_BYTES_LENGTH = 4;
+                    const [revertReasonHex] = ethers.utils.defaultAbiCoder.decode(
+                        ['bytes'],
+                        ethers.utils.hexDataSlice(returnData, REVERT_ERROR_SELECTOR_BYTES_LENGTH),
+                    );
```

Comment:
Re-use `decodeCallData` here with `Error(string)` as ABI to get `revertReasonHex`.

Reply:
It required some hacks because of the selector mismatch. But I've added an option to override the selector while decoding.